pyrdp/player/gdi/draw.py

```diff
@@ -0,0 +1,702 @@
+
+    def switchSurface(self, state: SwitchSurface):
+        LOG.debug(state)
+        if state.id not in self.surfaces:
+            # Appendix A - <5> Section 2.2.2.2.1.3.3
+            LOG.warning('Request for uninitialized surface: %d', state.id)
+            self.activeSurface = SUBSTITUTE_SURFACE
+            return
+        self.activeSurface = state.id
+
+    def createNineGridBitmap(self, state: CreateNineGridBitmap):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def streamBitmapFirst(self, state: StreamBitmapFirst):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def streamBitmapNext(self, state: StreamBitmapNext):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusFirst(self, state: GdiPlusFirst):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusNext(self, state: GdiPlusNext):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusEnd(self, state: GdiPlusEnd):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusCacheFirst(self, state: GdiPlusCacheFirst):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusCacheNext(self, state: GdiPlusCacheNext):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def drawGdiPlusCacheEnd(self, state: GdiPlusCacheEnd):
+        LOG.debug(state)
+        self._unimplemented()
+
+    def _process_glyph(self, ctx: 'GlyphContext', data: bytes):
+        """Process glyph rendering instructions."""
+        # 2.2.2.2.1.1.2.14
+        p = self._paint(self.surface)
+
+        cid = ctx.cacheId
+
+        i = 0
+        while i < len(data):
+            instr = data[i]
+            i += 1
+
+            if instr == GLYPH_FRAGMENT_USE:
+                fid = data[i]
+                i += 1
+                fragment = self.glyphs.getFragment(cid, fid)
+                n = 0
+                size = len(fragment)
+
+                while n < size:
+                    idx = fragment[n]
+                    n += 1
+
+                    glyph = self.glyphs.get(cid, idx)
+
+                    n = ctx.offset(n, fragment)
+                    ctx.draw(glyph, p)
+
+                # Skip for now, seems to always be 0.
+                if ctx.ulCharInc == 0 and not (ctx.flAccel & SO_CHAR_INC_EQUAL_BM_BASE):
+                    i += 2 if data[i] == 0x80 else 1
+
+            elif instr == GLYPH_FRAGMENT_ADD:
+                fid = data[i]
+                size = data[i + 1]
+                fragment = data[(i - size - 1):size]
+                i += 2
+                self.glyphs.addFragment(cid, fid, fragment)
+            else:
+                glyph = self.glyphs.get(cid, instr)
+                i = ctx.offset(i, data)
+                ctx.draw(glyph, p)
+        self._end(p)
+
+    def _unimplemented(self):
+        if not self._warned:
+            LOG.warning('One or more unimplemented drawing orders called! Expect lossy rendering.')
```

Comment:
:+1: 

Curious about how many hits we will have in the wild :smile:

Reply:
:D